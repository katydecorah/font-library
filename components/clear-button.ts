class ClearButton extends HTMLElement {
  constructor() {
    super();
  }
  connectedCallback() {
    // add aria-role
    this.setAttribute("role", "button");
    this.addEventListener("click", () => {
      this.dispatchEvent(
        new CustomEvent("clear-filter", {
          bubbles: true,
          composed: true,
          detail: {
            filter: this.getAttribute("value"),
          },
        })
      );
    });
  }
}

customElements.define("clear-button", ClearButton);
