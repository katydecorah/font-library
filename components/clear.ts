class ClearButton extends HTMLElement {
  constructor() {
    super();
  }
  connectedCallback() {
    this.addEventListener("click", () => {
      this.dispatchEvent(
        new CustomEvent("clear-filter", {
          bubbles: true,
          composed: true,
        })
      );
    });
  }
}

customElements.define("clear-button", ClearButton);
