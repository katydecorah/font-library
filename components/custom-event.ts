export default function customEvent(
  eventName: string,
  detail?: {
    value?: string | boolean | number;
    id?: string;
  }
): CustomEvent {
  return new CustomEvent(eventName, {
    bubbles: true,
    composed: true,
    detail,
  });
}
