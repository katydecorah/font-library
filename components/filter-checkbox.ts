import customEvent from "./custom-event";

class FilterCheckbox extends HTMLInputElement {
  constructor() {
    super();
    this.addEventListener("change", this.onChange);
    this.handleInitialValue();

    // Listen for events to clear filter
    window.addEventListener("remove-checkbox", () => {
      this.checked = false;
      this.onChange();
    });
  }

  onChange() {
    this.dispatchEvent(
      customEvent("handle-filter", {
        value: this.checked,
        id: this.id,
      })
    );
    this.setUrlParam();
  }

  setUrlParam() {
    const urlParameters = new URLSearchParams(window.location.search);
    // only set variable if it's true
    if (this.checked === false) {
      urlParameters.delete("variable");
    } else {
      urlParameters.set("variable", "true");
    }
    window.history.replaceState(
      {},
      "",
      `${window.location.pathname}?${urlParameters.toString()}`
    );
  }

  handleInitialValue() {
    const urlParameters = new URLSearchParams(window.location.search);
    const initialValue = urlParameters.get("variable");
    if (initialValue === "true") {
      this.checked = true;

      // Wait for main-app to load before dispatching event
      window.addEventListener("main-app-loaded", () => {
        this.onChange();
      });
    }
  }
}

customElements.define("filter-checkbox", FilterCheckbox, { extends: "input" });
