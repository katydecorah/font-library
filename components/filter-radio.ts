class FilterRadio extends HTMLInputElement {
  constructor() {
    super();
    this.onchange = this.onChange;

    // Get init value from URL param
    const urlParams = new URLSearchParams(window.location.search);
    const tag = urlParams.get("tag");
    if (tag === this.value) {
      this.checked = true;
    }
  }

  onChange() {
    this.dispatchEvent(
      new CustomEvent("tag-button-selected", {
        bubbles: true,
        composed: true,
        detail: {
          value: this.value,
        },
      })
    );

    this.setUrlParam();
  }

  setUrlParam() {
    const urlParams = new URLSearchParams(window.location.search);
    urlParams.set("tag", this.value.toString());
    window.history.replaceState(
      {},
      "",
      `${window.location.pathname}?${urlParams.toString()}`
    );
  }
}

customElements.define("filter-radio", FilterRadio, { extends: "input" });
