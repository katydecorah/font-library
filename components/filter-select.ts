class FilterSelect extends HTMLSelectElement {
  constructor() {
    super();
    this.onchange = this.onChange;

    // Get initial value from URL
    const urlParams = new URLSearchParams(window.location.search);
    const initValue = urlParams.get(this.dataset.param);
    if (initValue) {
      if (!this.options.namedItem(initValue)) return;
      this.value = initValue;

      // Wait for main-app to load before dispatching event
      window.addEventListener("main-app-loaded", () => {
        this.onChange();
      });
    }

    // Listen for events to clear filter
    window.addEventListener("remove-select", (e: CustomEvent) => {
      if (e.detail.filter === this.id) {
        this.value = "";
        this.onChange();
      }
    });

    // Listen for changes by other tag elements (radio, button)
    window.addEventListener("tag-button-selected", (e: CustomEvent) => {
      if (this.id === "selectedTag" && this.value !== e.detail.value) {
        this.value = e.detail.value;
      }
    });
  }

  onChange() {
    const eventName =
      this.id === "selectedTag" ? "tag-button-selected" : "handle-filter";
    this.dispatchEvent(
      new CustomEvent(eventName, {
        bubbles: true,
        composed: true,
        detail: { id: this.id, value: this.value },
      })
    );

    this.setUrlParam();
  }

  setUrlParam() {
    const param = this.dataset.param;
    const urlParams = new URLSearchParams(window.location.search);
    if (!this.value) {
      urlParams.delete(param);
    } else {
      urlParams.set(param, this.value.toString());
    }
    window.history.replaceState(
      {},
      "",
      `${window.location.pathname}?${urlParams.toString()}`
    );
  }
}

customElements.define("filter-select", FilterSelect, { extends: "select" });
