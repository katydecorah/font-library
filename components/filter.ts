import { GeneratedData } from "./main-app";

export default function filter(
  state: {
    selectedSearch: string;
    selectedTag: string;
    selectedCategory: string;
    selectedSubset: string;
    selectedVariant: string;
    selectedVariable: boolean;
    sortBy: string;
    currentPage: number;
    pageSize: number;
  },
  generatedData: GeneratedData
): [number, GeneratedData] {
  const {
    selectedSearch,
    selectedTag,
    selectedCategory,
    selectedSubset,
    selectedVariant,
    selectedVariable,
    sortBy,
    currentPage,
    pageSize,
  } = state;
  let filteredData = generatedData;

  if (selectedSearch) {
    filteredData = filteredData.filter((row) =>
      row.family.toLowerCase().includes(selectedSearch.toLowerCase())
    );
  }

  if (selectedTag && selectedTag !== "need tags") {
    filteredData = filteredData.filter((row) => row.tags.includes(selectedTag));
  }
  if (selectedTag === "need tags") {
    filteredData = filteredData.filter((row) => row.tags.length === 0);
  }
  if (selectedCategory) {
    filteredData = filteredData.filter(
      (row) => row.category === selectedCategory
    );
  }
  if (selectedSubset) {
    filteredData = filteredData.filter((row) =>
      row.subsets.includes(selectedSubset)
    );
  }
  if (selectedVariant) {
    filteredData = filteredData.filter((row) =>
      row.variants.includes(selectedVariant)
    );
  }

  if (selectedVariable === true) {
    filteredData = filteredData.filter((row) => row.variable);
  }

  if (sortBy === "date") {
    filteredData = filteredData.sort(
      (a, b) =>
        new Date(b.lastModified).getTime() - new Date(a.lastModified).getTime()
    );
  }

  if (sortBy === "family") {
    filteredData = filteredData.sort((a, b) =>
      a.family.localeCompare(b.family)
    );
  }

  const paginated = filteredData.filter((row, index) => {
    const start = (currentPage - 1) * pageSize;
    const end = currentPage * pageSize;
    if (index >= start && index < end) return true;
  });

  return [filteredData.length, paginated];
}
