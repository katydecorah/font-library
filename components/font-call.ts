import { GeneratedData } from "./font-results";

export default function fontCall({
  variants,
  slug,
  selectedVariant,
  previewName,
}: {
  variants: GeneratedData[number]["variants"];
  slug: string;
  selectedVariant: string;
  previewName: string;
}): string {
  let fontCallString = slug;

  if (selectedVariant && selectedVariant !== "regular") {
    fontCallString += fontCallSelectedVariant(selectedVariant);
  }
  // if font doesn't have regular variant, add subset to font call
  if (!selectedVariant && !variants.includes("regular")) {
    fontCallString += fontCallVariant(variants);
  }

  fontCallString += `&text=${encodeURIComponent(previewName)}&display=swap`;

  return `https://fonts.googleapis.com/css2?family=${fontCallString}`;
}

function fontCallVariant(variants: GeneratedData[number]["variants"]): string {
  const [firstVariant] = variants;
  if (/\d+/g.test(firstVariant)) {
    return `:wght@${firstVariant}`;
  } else if (firstVariant.includes("italic")) {
    return `:ital@1`;
  }
}

function fontCallSelectedVariant(selectedVariant: string): string {
  const hasItalic = selectedVariant.includes("italic");
  const variantNumber = selectedVariant.match(/\d+/g); // get number from selectedVariant

  const variants = [];
  if (selectedVariant === "italic") {
    variants.push("ital@1");
  } else if (hasItalic) {
    variants.push("ital");
  }
  if (variantNumber && variantNumber[0]) {
    variants.push(`wght@${hasItalic ? "1," : ""}${variantNumber[0]}`);
  }
  return `:${variants.join(",")}`;
}
