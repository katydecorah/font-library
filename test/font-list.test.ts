import "./components";

describe("FontList", () => {
  afterEach(() => {
    document.body.innerHTML = "";
  });

  it("should render with fonts", () => {
    document.body.innerHTML = `<ul is="font-list" fonts='[{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020},{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}]'></ul>`;
    const fontList = document.querySelector("ul[is='font-list']");
    expect(fontList).toMatchSnapshot();
  });

  it("should render with fonts and selectedVariant", () => {
    document.body.innerHTML = `<ul is="font-list" fonts='[{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020},{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}]' selected-variant="700"></ul>`;
    const fontList = document.querySelector("ul[is='font-list']");
    expect(fontList).toMatchSnapshot();
  });

  it("should render with fonts and selectedSubset", () => {
    document.body.innerHTML = `<ul is="font-list" fonts='[{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020},{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}]' selected-subset="hebrew"></ul>`;
    const fontList = document.querySelector("ul[is='font-list']");
    expect(fontList).toMatchSnapshot();
  });
});
