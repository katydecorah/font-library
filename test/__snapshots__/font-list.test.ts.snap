// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FontList should render with fonts 1`] = `
<ul
  fonts="[{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020},{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}]"
  is="font-list"
>
  <li
    font="{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020}"
    is="font-item"
  >
    <div
      class="family-link"
    >
      
      
      <div
        class="family-title unifrakturcook"
        style="font-family: 'UnifrakturCook';"
      >
        
        UnifrakturCook
      
      </div>
      
      
      <div
        class="family-meta-container"
      >
        
        
        <span
          class="family-title-small"
        />
        
        
        <div
          class="family-meta"
        >
          
          
          <ul>
            
            
            <li>
              display
            </li>
            
            
            <li>
              <span
                aria-label="700"
              >
                1 variants
              </span>
            </li>
            
            
            <li>
              <span
                aria-label="latin"
              >
                1 subsets
              </span>
            </li>
            
          
          </ul>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="family-tags"
    >
      
      
      <div
        class="family-tags-container"
      />
      
      
      <div
        class="family-meta-links"
      >
        
        
        <a
          aria-label="Edit tags for UnifrakturCook"
          href="https://github.com/katydecorah/font-library/blob/gh-pages/families.json#L1020"
          target="_blank"
        >
          Edit tags
        </a>
        
        
        <a
          aria-label="Visit UnifrakturCook on Google Fonts"
          href="https://fonts.google.com/specimen/UnifrakturCook"
          target="_blank"
        >
          Google Fonts →
        </a>
        
      
      </div>
      
    
    </div>
  </li>
  

  <li
    font="{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}"
    is="font-item"
  >
    <div
      class="family-link"
    >
      
      
      <div
        class="family-title alef"
        style="font-family: 'Alef';"
      >
        
        Alef
      
      </div>
      
      
      <div
        class="family-meta-container"
      >
        
        
        <span
          class="family-title-small"
        />
        
        
        <div
          class="family-meta"
        >
          
          
          <ul>
            
            
            <li>
              sans-serif
            </li>
            
            
            <li>
              <span
                aria-label="regular, 700"
              >
                2 variants
              </span>
            </li>
            
            
            <li>
              <span
                aria-label="hebrew, latin"
              >
                2 subsets
              </span>
            </li>
            
          
          </ul>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="family-tags"
    >
      
      
      <div
        class="family-tags-container"
      >
        <button
          is="tag-button"
          style="font-family: "Alatsi";"
          title="geometric"
          value="geometric"
        >
          geometric
        </button>
      </div>
      
      
      <div
        class="family-meta-links"
      >
        
        
        <a
          aria-label="Edit tags for Alef"
          href="https://github.com/katydecorah/font-library/blob/gh-pages/families.json#L23"
          target="_blank"
        >
          Edit tags
        </a>
        
        
        <a
          aria-label="Visit Alef on Google Fonts"
          href="https://fonts.google.com/specimen/Alef"
          target="_blank"
        >
          Google Fonts →
        </a>
        
      
      </div>
      
    
    </div>
  </li>
</ul>
`;

exports[`FontList should render with fonts and selectedSubset 1`] = `
<ul
  fonts="[{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020},{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}]"
  is="font-list"
  selected-subset="hebrew"
>
  <li
    font="{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020}"
    is="font-item"
    selected-subset="hebrew"
  >
    <div
      class="family-link"
    >
      
      
      <div
        class="family-title unifrakturcook"
        style="font-family: 'UnifrakturCook';direction: rtl;"
      >
        
        שָׁלוֹם עוֹלָם 
      
      </div>
      
      
      <div
        class="family-meta-container"
      >
        
        
        <span
          class="family-title-small"
        >
          UnifrakturCook
        </span>
        
        
        <div
          class="family-meta"
        >
          
          
          <ul>
            
            
            <li>
              display
            </li>
            
            
            <li>
              <span
                aria-label="700"
              >
                1 variants
              </span>
            </li>
            
            
            <li>
              <span
                aria-label="latin"
              >
                1 subsets
              </span>
            </li>
            
          
          </ul>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="family-tags"
    >
      
      
      <div
        class="family-tags-container"
      />
      
      
      <div
        class="family-meta-links"
      >
        
        
        <a
          aria-label="Edit tags for UnifrakturCook"
          href="https://github.com/katydecorah/font-library/blob/gh-pages/families.json#L1020"
          target="_blank"
        >
          Edit tags
        </a>
        
        
        <a
          aria-label="Visit UnifrakturCook on Google Fonts"
          href="https://fonts.google.com/specimen/UnifrakturCook"
          target="_blank"
        >
          Google Fonts →
        </a>
        
      
      </div>
      
    
    </div>
  </li>
  

  <li
    font="{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}"
    is="font-item"
    selected-subset="hebrew"
  >
    <div
      class="family-link"
    >
      
      
      <div
        class="family-title alef"
        style="font-family: 'Alef';direction: rtl;"
      >
        
        שָׁלוֹם עוֹלָם 
      
      </div>
      
      
      <div
        class="family-meta-container"
      >
        
        
        <span
          class="family-title-small"
        >
          Alef
        </span>
        
        
        <div
          class="family-meta"
        >
          
          
          <ul>
            
            
            <li>
              sans-serif
            </li>
            
            
            <li>
              <span
                aria-label="regular, 700"
              >
                2 variants
              </span>
            </li>
            
            
            <li>
              <span
                aria-label="hebrew, latin"
              >
                2 subsets
              </span>
            </li>
            
          
          </ul>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="family-tags"
    >
      
      
      <div
        class="family-tags-container"
      >
        <button
          is="tag-button"
          style="font-family: "Alatsi";"
          title="geometric"
          value="geometric"
        >
          geometric
        </button>
      </div>
      
      
      <div
        class="family-meta-links"
      >
        
        
        <a
          aria-label="Edit tags for Alef"
          href="https://github.com/katydecorah/font-library/blob/gh-pages/families.json#L23"
          target="_blank"
        >
          Edit tags
        </a>
        
        
        <a
          aria-label="Visit Alef on Google Fonts"
          href="https://fonts.google.com/specimen/Alef"
          target="_blank"
        >
          Google Fonts →
        </a>
        
      
      </div>
      
    
    </div>
  </li>
</ul>
`;

exports[`FontList should render with fonts and selectedVariant 1`] = `
<ul
  fonts="[{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020},{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}]"
  is="font-list"
  selected-variant="700"
>
  <li
    font="{"family":"UnifrakturCook","variants":["700"],"subsets":["latin"],"category":"display","tags":[],"lineNumber":1020}"
    is="font-item"
    selected-variant="700"
  >
    <div
      class="family-link"
    >
      
      
      <div
        class="family-title unifrakturcook"
        style="font-family: 'UnifrakturCook';"
      >
        
        UnifrakturCook
      
      </div>
      
      
      <div
        class="family-meta-container"
      >
        
        
        <span
          class="family-title-small"
        />
        
        
        <div
          class="family-meta"
        >
          
          
          <ul>
            
            
            <li>
              display
            </li>
            
            
            <li>
              <span
                aria-label="700"
              >
                1 variants
              </span>
            </li>
            
            
            <li>
              <span
                aria-label="latin"
              >
                1 subsets
              </span>
            </li>
            
          
          </ul>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="family-tags"
    >
      
      
      <div
        class="family-tags-container"
      />
      
      
      <div
        class="family-meta-links"
      >
        
        
        <a
          aria-label="Edit tags for UnifrakturCook"
          href="https://github.com/katydecorah/font-library/blob/gh-pages/families.json#L1020"
          target="_blank"
        >
          Edit tags
        </a>
        
        
        <a
          aria-label="Visit UnifrakturCook on Google Fonts"
          href="https://fonts.google.com/specimen/UnifrakturCook"
          target="_blank"
        >
          Google Fonts →
        </a>
        
      
      </div>
      
    
    </div>
  </li>
  

  <li
    font="{"family":"Alef","variants":["regular","700"],"subsets":["hebrew","latin"],"category":"sans-serif","tags":["geometric"],"lineNumber":23}"
    is="font-item"
    selected-variant="700"
  >
    <div
      class="family-link"
    >
      
      
      <div
        class="family-title alef"
        style="font-family: 'Alef';"
      >
        
        Alef
      
      </div>
      
      
      <div
        class="family-meta-container"
      >
        
        
        <span
          class="family-title-small"
        />
        
        
        <div
          class="family-meta"
        >
          
          
          <ul>
            
            
            <li>
              sans-serif
            </li>
            
            
            <li>
              <span
                aria-label="regular, 700"
              >
                2 variants
              </span>
            </li>
            
            
            <li>
              <span
                aria-label="hebrew, latin"
              >
                2 subsets
              </span>
            </li>
            
          
          </ul>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="family-tags"
    >
      
      
      <div
        class="family-tags-container"
      >
        <button
          is="tag-button"
          style="font-family: "Alatsi";"
          title="geometric"
          value="geometric"
        >
          geometric
        </button>
      </div>
      
      
      <div
        class="family-meta-links"
      >
        
        
        <a
          aria-label="Edit tags for Alef"
          href="https://github.com/katydecorah/font-library/blob/gh-pages/families.json#L23"
          target="_blank"
        >
          Edit tags
        </a>
        
        
        <a
          aria-label="Visit Alef on Google Fonts"
          href="https://fonts.google.com/specimen/Alef"
          target="_blank"
        >
          Google Fonts →
        </a>
        
      
      </div>
      
    
    </div>
  </li>
</ul>
`;

exports[`FontList should return if no fonts 1`] = `
<ul
  is="font-list"
/>
`;
